### PR TITLE
Fix #2990

### LINKED ISSUE #2990: linkcheck.py is broken?
Regarding to #2988: in the first place, linkcheck.py is broken now?

There are at least two problems.
1. for Python 3.x (for me, it is 3.5.2), linkcheck.py is likely to have a unicode problem.
2. for both 3.x and 2.7, [#L167](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/builders/linkcheck.py#L167) may raise `UnboundLocalError` (`local variable 'code' referenced before assignment`).

For now, running `make linkcheck` in CPython documentation fails, because the UnboundLocalError occurs.

Former is like this:

```
distributing\index.rst:123: [broken] https://packaging.python.org/en/latest/distributing/#packaging-your-project: Can't convert 'bytes' object to str implicitly
distributing\index.rst:124: [broken] https://packaging.python.org/en/latest/distributing/#uploading-your-project-to-pypi: Can't convert 'bytes' object to str implicitly
distributing\index.rst:111: [broken] https://packaging.python.org/en/latest/current/#packaging-tool-recommendations: Can't convert 'bytes' object to str implicitly
distutils\apiref.rst:1002: [broken] http://nfs.sourceforge.net/#section_d: Can't convert 'bytes' object to str implicitly
howto\descriptor.rst:126: [broken] https://www.python.org/download/releases/2.2.3/descrintro/#cooperation: Can't convert 'bytes' object to str implicitly
howto\logging-cookbook.rst:1639: [broken] https://tools.ietf.org/html/rfc5424#section-6: Can't convert 'bytes' object to str implicitly
howto\logging-cookbook.rst:1376: [broken] https://docs.djangoproject.com/en/1.9/topics/logging/#configuring-logging: Can't convert 'bytes' object to str implicitly
howto\logging-cookbook.rst:1320: [broken] https://docs.djangoproject.com/en/1.9/topics/logging/#configuring-logging: Can't convert 'bytes' object to str implicitly
...
```

It is likely that all of these has a fragment...

### PR DESCRIPTION
**Pull request by hhsprings**
For #2990 .

I believe this PR is proper fix, but I don't know "why does this occur" for `redirect but 200 OK', so please add rationale as comment, or review comment.

I added test_build_linkcheck.py but this is very minimal for future any fixes (for example #2988). I think maintainance of `test_build_linkcheck.py` would be too heavy, so I could not decide writing rich testing.

### CODE REVIEW

File: tests/roots/test-linkcheck/conf.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+master_doc = 'links'
+source_suffix = '.txt'
+exclude_patterns = ['_build']
```
Comment (tk0miya) [inline tests/roots/test-linkcheck/conf.py:3]: In my understanding, #2990 is happened with `linkcheck_anchors` option. But it seems `test-linkcheck` project does not define it. Does this really test the problem?
If you remove the `decode_unicode` parameter, the problem will be reproduced by this?
Could you check that please?

File: tests/roots/test-linkcheck/links.txt [hunk 2]
```diff
@@ -0,0 +1,167 @@
+This is from CPython documentation.
```
Comment (tk0miya) [inline tests/roots/test-linkcheck/links.txt:1]: The whole of document is needed to test for #2990?
This test will be executed thousands of times by developers and Travis.
If such a large document is not important for test, please shrink this please.

**Review by tk0miya — CHANGES_REQUESTED** [review]

**Comment by hhsprings** [comment]
Thank you for reviewing. I'd shrunk links.txt, and added proper assert in test_build_linkcheck.py.

**Review by tk0miya — APPROVED** [review]
Thank you for updating. LGTM!

**Comment by tk0miya** [comment]
Merged. Thank you for your contribution!

**Pull request merged**